File: leetcode/searchInsert/searchInsert.py
# Given a sorted array and a target value, return the index if the target is found. 
# If not, return the index where it would be if it were inserted in order.

# You may assume no duplicates in the array.

# Example 1:
# Input: [1,3,5,6], 5
# Output: 2

# Example 2:
# Input: [1,3,5,6], 2
# Output: 1

# Example 3:
# Input: [1,3,5,6], 7
# Output: 4

# Example 4:
# Input: [1,3,5,6], 0
# Output: 0


class Solution:
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        beg = 0
        end = len(nums)
        mid = beg + (end-beg) // 2
        while(beg != end and nums[mid] != target):
            if nums[mid] < target:
                beg = mid + 1
            else:
                end = mid
            mid = beg + (end - beg) // 2
        return mid


if __name__ == '__main__':
   solver = Solution()
   print(solver.searchInsert([1,3,5,6], 5))
   print(solver.searchInsert([1,3,5,6], 2))
   print(solver.searchInsert([1,3,5,6], 7))
   print(solver.searchInsert([1,3,5,6], 0))

   

File: leetcode/searchInsert/searchInsert.cpp
#include <vector>
#include <iostream>
using namespace std;


class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        auto beg = nums.begin(), end = nums.end();
        auto mid = nums.begin() + (nums.end() - nums.begin())/ 2;
        while(beg != end && *mid != target){
            if (target < *mid)
                end = mid;
            else    
                beg = mid + 1;
            mid = beg + (end - beg) / 2;
        }
        return mid - nums.begin();
    }
};


int main(){
    Solution solver;
    vector<int> nums{1,3,5,6};
    cout << solver.searchInsert(nums, 5) << endl;
    cout << solver.searchInsert(nums, 2) << endl;
    cout << solver.searchInsert(nums, 7) << endl;
    cout << solver.searchInsert(nums, 0) << endl;
    return 0;
}